doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,396 @@ Applying theorems
+     :n:`one_term` (substituting for the unified variables)
+     and the tactic creates a new subgoal for each unmatched premise.
+     See the example :ref:`here <apply_forward>`.
+
+     If specified, :n:`as @simple_intropattern` is applied to the conclusion
+     of the type of :n:`@one_term`.
+
+     If the type of :n:`@one_term` is an inductive type isomorphic to a tuple type,
+     then the tuple is (recursively) decomposed and the first component of the
+     tuple for which a non-dependent premise matches the conclusion of the
+     hypothesis is used.  :ref:`Example <example_apply_hyp_inductive>`
+
+     For the hypothesis case, matching is done only with first-order unification.
+
+   :n:`with @bindings` (in :n:`one_term_with_bindings`)
+     Gives explicit instantiations for variables used in the type of :n:`@one_term`.
+     There are 3 cases:
+
+     - Bindings for variables can be provided in a list of :n:`@one_term`\s
+       in the left-to-right order of their first appearance in the type of
+       :n:`@one_term`.  For the goal case (:ref:`example <apply_with_binding_goal>`),
+       the list should give bindings only for variables that aren't bound by
+       unification.  However, in the hypothesis case
+       (:ref:`example <apply_with_binding_hyp>`),
+       the list must include bindings for *all* variables.
+
+     - Bindings for unbound variables can be given by name with the
+       :n:`(@ident := @term)` form.
+
+     - The form :n:`(@natural := @term)` binds additional variables by
+       unifying the Nth non-dependent premise of the type of :n:`@one_term` with :n:`@term`.
+       (Use `1` for the first premise.)
+
+   .. exn:: Unable to unify @one_term with @one_term.
+
+      The :tacn:`apply` tactic failed to match the conclusion of :token:`one_term`.
+      You can help :tacn:`apply` by
       transforming your goal with the :tacn:`change` or :tacn:`pattern`
       tactics.
 
+   .. exn:: Unable to apply lemma of type "..." on hypothesis of type "...".
+
+      This happens if the conclusion of :token:`ident` does not match any of
+      the non-dependent premises of the type of :token:`term`.
+
    .. exn:: Unable to find an instance for the variables {+ @ident}.
 
-      This occurs when some instantiations of the premises of :token:`term` are not deducible
+      This occurs when some instantiations of the premises of :token:`one_term` are not deducible
       from the unification. This is the case, for instance, when you want to apply a
-      transitivity property. In this case, you have to use one of the variants below:
+      transitivity property.  To fix this, add bindings for the :n:`@ident`\s using
+      to :n:`with @bindings` or use :tacn:`eapply`.
 
-   .. tacv:: apply @term with {+ @term}
+   .. todo: we should be listing things like "Debug tactic-unification" in
+      in the options index.  Maybe we should add ":debug:" as a new tag.  WDYT?
```

Comment:
Is this supposed to be a `todo PR`? If not, then maybe remove the WDYT. (I think I already gave my opinion that this was a good idea but probably not in scope for this PR.)